lazy-pages/src/lib.rs

```diff
@@ -20,68 +20,142 @@
+    /// Found a signal from same page twice - see more in head comment.
+    #[display(fmt = "Page cannot be release page twice: {:?}", _0)]
+    DoubleRelease(PageNumber),
+    #[display(fmt = "Protection error: {}", _0)]
+    #[from]
+    MemoryProtection(region::Error),
+}
 
-    /// Identify whether signal handler is set for current thread
-    static LAZY_PAGES_ENABLED: RefCell<bool> = RefCell::new(false);
+#[derive(Default)]
+pub(crate) struct LazyPagesExecutionContext {
     /// Pointer to the begin of wasm memory buffer
-    static WASM_MEM_BEGIN: RefCell<HostPointer> = RefCell::new(0);
-    /// Key in storage for each lazy page
-    static LAZY_PAGES_INFO: RefCell<BTreeMap<LazyPage, Vec<u8>>> = RefCell::new(BTreeMap::new());
+    pub wasm_mem_addr: Option<HostPointer>,
+    /// Current program prefix in storage
+    pub program_storage_prefix: Option<Vec<u8>>,
+    /// Set of lazy pages
+    pub lazy_pages: BTreeSet<PageNumber>,
     /// Page data, which has been in storage before current execution.
     /// For each lazy page, which has been accessed.
-    static RELEASED_LAZY_PAGES: RefCell<BTreeMap<LazyPage, Option<PageBuf>>> = RefCell::new(BTreeMap::new());
+    pub released_lazy_pages: BTreeMap<PageNumber, Option<PageBuf>>,
+
+    #[deprecated]
+    /// Keys in storage for each lazy page.
+    pub lazy_pages_info: BTreeMap<PageNumber, Vec<u8>>,
+}
+
+thread_local! {
+    // NOTE: here we suppose, that each contract is executed in separate thread.
+    // Or may be in one thread but consequentially.
+
+    /// Identify whether signal handler is set for current thread
+    static LAZY_PAGES_ENABLED: RefCell<bool> = RefCell::new(false);
+    /// Lazy pages context for current execution
+    static LAZY_PAGES_CONTEXT: RefCell<LazyPagesExecutionContext> = RefCell::new(Default::default());
 }
 
-pub fn save_lazy_pages_info(pages: Vec<u32>, prefix: Vec<u8>) {
-    LAZY_PAGES_INFO.with(|lazy_pages_info| {
-        let mut lazy_pages_info = lazy_pages_info.borrow_mut();
-        pages.into_iter().for_each(|p| {
-            let mut key = Vec::with_capacity(prefix.len() + std::mem::size_of::<u32>());
-            key.extend(prefix.clone());
-            key.extend(p.to_le_bytes().to_vec());
-            lazy_pages_info.insert(LazyPage(p), key);
-        })
+/// Set lazy pages numbers and current program prefix in global context
+pub fn set_lazy_pages_info(pages: Vec<u32>, prefix: Vec<u8>) {
+    LAZY_PAGES_CONTEXT.with(|ctx| {
+        let mut ctx = ctx.borrow_mut();
+        ctx.lazy_pages
+            .extend(pages.into_iter().map(PageNumber::from));
+        ctx.program_storage_prefix.replace(prefix);
     });
 }
 
-/// Returns vec of not-accessed wasm lazy pages
-pub fn available_pages() -> Vec<LazyPage> {
-    LAZY_PAGES_INFO.with(|lazy_pages_info| lazy_pages_info.borrow().keys().copied().collect())
+/// Returns lazy pages numbers (not accessed yet)
+pub fn get_lazy_pages() -> Vec<PageNumber> {
```

Comment:
Why was it renamed?

Reply:
old name is counterintuitively

Reply:
Can you explain in more detail?

Reply:
Is now deprecated - so, no reasons to discuss this